# Inferential Statistics
menggunakan statistik untuk membuat konklusi

Terminologi:
- random variable
    - discrete
    - continuous

Konsep:
- Probability Distribution
- Confidence Interval
- Hypothesis Testing

setiap titik data (numerik) yang kita kumpulkan/ survey/ yang ada di raw data adalah sebuah random variable. random variable artinya adalah nilai yang sifatnya acak, yang tidak kita ketahui.

Ada dua jenis kategori data random numerik ini:
- discrete, bulat, tidak ada diantara, seperti hasil lempar dadu, value hanya ada di titik tertentu
- continuous, tidak ada batasan. bisa ada berapa banyak pun value antara dua titik. contoh: berat badan, antara yang 2 - 3kg, bisa saja ada 2.01kg 2.53kg 2.715kg tingkat desimal poin hanya terbatas oleh instrumen yang digunakan.

## Probability Distribution
sebuah probability distribution adalah
- daftar kemungkinan masing masing value (untuk discrete random variable)
- daftar kemungkinan satu grup value (untuk continuous random variable)

# Confidence Interval
membuat estimasi dari data yang diamati, kita membuat range rata-rata bukan 1 angka rata-rata. 

cara hitung:
1. Best Estimate +- Margin of Error

misalkan best estimate kita adalah 10 dan margin of error adalah 2, maka confidence inverval adalah antara 8 dan 12: [8, 12]

2. Best Estimate? mean, atau proportion (persentase)

3. Margin of error? z * estimated standard error

4. z? diatur dari tingkat keyakinan confidence interval kita, normalnya z ada di 1.96 untuk 95% keyakinan

5. estimated standard error?
    - proporsi: `np.sqrt( p * (1-p) / n )`
    - mean: `stddev / np.sqrt(n)`

6. p? --> tentukan % apa yang mau dicari, misalkan jumlah orang yang beratnya 50-60kg, ternyata ad 5 dari 12 orang yg di survey, maka p = 5/12
7. mean? stddev? n? --> diambil langsung dari raw data

In [28]:
# import package
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats

import statsmodels.api as sm

In [2]:
# import data
df_heart = pd.read_csv("../dataset/heart.csv")
df_heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [3]:
df_heart.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [4]:
df_heart.head() 

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

### Buatlah Confidence Interval untuk proporsi pria yang sakit jantung

di data ini, 
- kolom target artinya (0 = sehat; 1 = sakit jantung)
- kolom sex artinya (1 = male; 0 = female)

In [5]:
df_heart["sex"].value_counts()

1    713
0    312
Name: sex, dtype: int64

ada 713 pria. dari 713 pria ada berapa yang sakit jantung (target = 1)?

In [7]:
df_heart_pria = df_heart[df_heart["sex"] == 1]
df_heart_pria_sakit = df_heart_pria[df_heart_pria["target"] == 1]

print(len(df_heart_pria), len(df_heart_pria_sakit))

713 300


dari 713 pria ada 300 yang sakit jantung

In [8]:
n = 713
p = 300 / 713

se = np.sqrt( p * (1-p) / n )

z = 1.96

me = z * se

be = p

print([be - me, be + me])

[0.3845199298886166, 0.45699479661909725]


38 - 46 % pria mengidap sakit jantung

### coba cari Confidence Interval untuk para wanita yang sehat

In [12]:
df_wanita = df_heart[df_heart["sex"] == 0]
df_wanita_sehat = df_wanita[df_wanita["target"] == 0]

n = len(df_wanita)
p = len(df_wanita_sehat) / n

se = np.sqrt( p * (1-p) / n )
z = 1.96
me = z * se
be = p

print([be - me, be + me])

[0.22605861640768202, 0.32522343487436933]


22.6 - 32.5 % wanita sehat tanpa penyakit jantung

## Confidence Interval untuk Mean
- coba cari rata2 kolesterol untuk orang yang punya penyakit jantung
- coba cari rata2 kolesterol untuk orang yang TIDAK punya penyakit jantung

In [23]:
df_sakit = df_heart[df_heart["target"] == 1]

mean = df_sakit["chol"].mean()
stddev = df_sakit["chol"].std()
n = len(df_sakit)

print(mean, stddev, n)

se = stddev / np.sqrt(n)
z = 1.96
me = z * se
be = mean

print(be - me, be + me)

240.97908745247148 53.010345260991095 526
236.4488201044286 245.50935480051436


rata2 kolesterol orang yang punya penyakit jantung ada di 236 - 246

In [24]:
# cari confidence interval untuk rata rata tingkat kolesterol orang yang TIDAK punya penyakit
df_sehat = df_heart[df_heart["target"] == 0]

mean = df_sehat["chol"].mean()
stddev = df_sehat["chol"].std()
n = len(df_sehat)

print(mean, stddev, n)

se = stddev / np.sqrt(n)
z = 1.96
me = z * se
be = mean

print(be - me, be + me)

251.2925851703407 49.55892364429828 499
246.9442034235395 255.64096691714187


rata2 kolesterol orang yang TIDAK punya penyakit jantung ada di 247-256

# Hypothesis Testing
teknik untuk membuat sebuah pendapat dari data. Setelah pendapat itu dibuat, kita berikan nama "Alternate Hypothesis H1" lalu kita buat statement yang menentang pendapat tersebut dan diberi nama "Null Hypothesis H0" sebagai asumsi awal. Kita kemudian mencari p-value untuk menunjukkan jika Null Hypothesis itu kita REJECT atau tidak.

- one population proportion
    - punya 1 set data
    - dibandingkan dengan 1 proporsi

- two population means
    - punya 2 set data
    - membandingkan rata2 dari kedua set data tersebut

setelah mendapat p-value, kita bandingkan p-value dengan error rate kita. dengan tingkat keyakinan di 95% artinya error rate kita 5%, alpha = 0.05.

jika p-value < alpha maka kita REJECT null hypothesis

jika p-value >= alpha, maka kita FAIL TO REJECT null hypothesis

In [27]:
import statsmodels.api as sm

dulu 52% orangtua percaya x. sekarang lebih banyak, iya tidak?

- H0: p <= 0.52 (tidak, sekarang sama saja jumlah orang yg percaya atau bahkan lebih sedikit)
- H1: p > 0.52  (iya sekarang lebih banyak yang percaya)

data yang kita punya, survei membuktikan: 1018 orangtua, 56% percaya x.

In [31]:
n = 1018
pnull = 0.52
phat = 0.56

zstat, pvalue = sm.stats.proportions_ztest(phat * n, n, pnull, alternative='larger')

alpha = 0.05 # 95% yakin berarti 5% error rate

print("p-value", pvalue)
if (pvalue < alpha):
    print("REJECT null hypothesis H0")
else:
    print("FAIL TO REJECT null hypothesis H0")

p-value 0.005069273865860533
REJECT null hypothesis H0


karena kita berhasil REJECT null hypothesis, maka kita bisa bilang kalau iya, sekarang lebih banyak orangtua yang percaya x. Ada perbedaan signifikan antara proporsi orangtua yang percaya x dulu dengan sekarang.

Note: untuk pilihan alternative ada smaller, two-sided, dan larger. 
- `smaller` adalah saat H1 mau membuktikan sekarang lebih kecil. 
- `two-sided` adalah kalau H1 mau membuktikan sekarang dan dulu berbeda. 
- terakhir `larger` adalah saat H1 mau membuktikan sekarang lebih besar.

link ke dokumentasi untuk statsmodels proportions_ztest https://www.statsmodels.org/dev/generated/statsmodels.stats.proportion.proportions_ztest.html

### Difference in Population Means
perbedaan rata-rata antara dua set data

contoh kali ini akan menggunakan data NHANES dari https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nhanes_2015_2016.csv tentang survey kesehatan orang Amerika

In [32]:
df_nhanes = pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nhanes_2015_2016.csv')
df_nhanes.head()

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
3       1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4       1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  
3     37.7     38.3     110.1     2.0  
4     36.0     27.2      80.4     2.0  

[5 rows x 28 columns]

apakah rata2 BMI wanita dan pria berbeda?

- kolom RIAGENDR 1 pria, 2 wanita
- kolom BMXBMI menunjukkan BMI seseorang

pembuatan hypothesis (null dan alternate)
- H0: bmi_p == bmi_w
- H1: bmi_p != bmi_w

kemudian kita cari p-value

In [38]:
df_p = df_nhanes[df_nhanes["RIAGENDR"] == 1]
df_w = df_nhanes[df_nhanes["RIAGENDR"] == 2]

df_bmi_p = df_p["BMXBMI"]
df_bmi_w = df_w["BMXBMI"]

# hapus missing value
df_bmi_p = df_bmi_p.dropna()
df_bmi_w = df_bmi_w.dropna()

# cek2 rata2, std dev, total data
print("pria")
print(df_bmi_p.agg(['mean', 'std', 'count']))
print()

print("wanita")
print(df_bmi_w.agg(['mean', 'std', 'count']))

pria
mean       28.778072
std         6.252568
count    2718.000000
Name: BMXBMI, dtype: float64

wanita
mean       29.939946
std         7.753319
count    2944.000000
Name: BMXBMI, dtype: float64


In [39]:
zstat, pvalue = sm.stats.ztest(df_bmi_p, df_bmi_w, alternative="two-sided")

alpha = 0.05 # 95% keyakinan 5% error rate jadi alpha 0.05

print("p-value", pvalue)
if (pvalue < alpha):
    print("REJECT null hypothesis H0")
else:
    print("FAIL TO REJECT null hypothesis H0")

p-value 6.591544431126401e-10
REJECT null hypothesis H0


rata2 BMI pria dan wanita berbeda

## latihan hypothesis testing
gunakan data df_heart, apakah rata2 kolesterol wanita lebih tinggi daripada rata2 kolesterol pria?

- membuat H1
- membuat H0
- mencari p-value
- set alpha
- REJECT atau FAIL TO REJECT null hypothesis H0
- konklusi

#### rata2 dua populasi
- H0: chol_w <= chol_p
- H1: chol_w > chol_p


In [44]:
df_wanita = df_heart[df_heart["sex"] == 0]
df_pria = df_heart[df_heart["sex"] == 1]

df_chol_w = df_wanita["chol"].dropna()
df_chol_p = df_pria["chol"].dropna()

print("wanita", df_chol_w.mean(), df_chol_w.std(), len(df_chol_w))
print("pria", df_chol_p.mean(), df_chol_p.std(), len(df_chol_p))

wanita 261.45512820512823 64.46678079217995 312
pria 239.23702664796633 43.15553498840588 713


In [45]:
zstat, pvalue = sm.stats.ztest(df_chol_w, df_chol_p, alternative="larger")

alpha = 0.05

print(pvalue)
if (pvalue < alpha):
    print("REJECT null hypothesis H0")
else:
    print("FAIL TO REJECT null hypothesis H0")

4.913943666354449e-11
REJECT null hypothesis H0


rata2 kolesterol wanita lebih tinggi daripada rata2 kolesterol pria